yesod-test/Yesod/Test.hs

```diff
@@ -397,13 +424,8 @@ yesodSpecWithSiteGeneratorAndArgument getSiteAction yspecs =
       unYesod getSiteAction' (YesodSpecGroup x y) = Hspec.specGroup x $ map (unYesod getSiteAction') y
       unYesod getSiteAction' (YesodSpecItem x y) = Hspec.specItem x $ \a -> do
         site <- getSiteAction' a
-        app <- toWaiAppPlain site
-        evalSIO y YesodExampleData
-            { yedApp = app
-            , yedSite = site
-            , yedCookies = M.empty
-            , yedResponse = Nothing
-            }
+        yed <- yesodExampleDataFromApp site
+        evalSIO y yed
```

Comment:
So, a problem with the current implementation (as surfaced in the pending test), is that we can't persist state from the `YesodExampleData` that is generated. We're doing `evalSIO y yed` and throwing away the modified state.

This means our `beforeApp` and `beforeWithApp` hooks can modify the database and return database entities, but those entities (and any changes made to the app, like sessions or logins) won't be kept.

A proper fix would probably thread the `YesodExampleData` through the whole thing, and then only let it get cleared out in the `Example` instance.

Reply:
I haven't looked in depth yet, but would using a mutable variable be a solution here?

Reply:
The `SIO` stuff is already a mutable variable. We're just not threading it through and re-using it afterwards.